crates/core/src/common.rs

```diff
@@ -0,0 +1,218 @@
+use kitsune2_api::{
+    AgentInfoSigned, DhtArc, DynLocalAgentStore, DynPeerStore, K2Result,
+};
+use std::collections::HashSet;
+use std::sync::Arc;
+
+/// Get all remote agents from the peer store.
+///
+/// Uses the [DynPeerStore] to get all agents and the [DynLocalAgentStore] to filter out local
+/// agents.
+pub async fn get_all_remote_agents(
+    peer_store: DynPeerStore,
+    local_agent_store: DynLocalAgentStore,
+) -> K2Result<Vec<Arc<AgentInfoSigned>>> {
+    let all_agents = peer_store.get_all().await?;
+    remove_local_agents(local_agent_store, all_agents).await
+}
+
+/// Get remote agents by overlapping storage arc.
+///
+/// Uses the [DynPeerStore] to get agents with overlapping storage arcs and the [DynLocalAgentStore]
+/// to filter out local agents.
+pub async fn get_remote_agents_by_overlapping_storage_arc(
+    peer_store: DynPeerStore,
+    local_agent_store: DynLocalAgentStore,
+    arc: DhtArc,
+) -> K2Result<Vec<Arc<AgentInfoSigned>>> {
+    let by_overlapping_storage_arc =
+        peer_store.get_by_overlapping_storage_arc(arc).await?;
+    remove_local_agents(local_agent_store, by_overlapping_storage_arc).await
+}
+
+/// Get remote agents near a specific location.
+///
+/// Uses the [DynPeerStore] to get agents near a location and the [DynLocalAgentStore] to filter
+/// out local agents.
+pub async fn get_remote_agents_near_location(
+    peer_store: DynPeerStore,
+    local_agent_store: DynLocalAgentStore,
+    loc: u32,
+    limit: usize,
+) -> K2Result<Vec<Arc<AgentInfoSigned>>> {
+    let near_location = peer_store.get_near_location(loc, limit).await?;
+    remove_local_agents(local_agent_store, near_location).await
+}
+
+async fn remove_local_agents(
+    local_agent_store: DynLocalAgentStore,
+    agent_list: Vec<Arc<AgentInfoSigned>>,
+) -> K2Result<Vec<Arc<AgentInfoSigned>>> {
+    let all_local_agents = local_agent_store
+        .get_all()
+        .await?
+        .into_iter()
+        .map(|a| a.agent().clone())
+        .collect::<HashSet<_>>();
+
+    Ok(agent_list
+        .into_iter()
+        .filter(|a| !all_local_agents.contains(&a.agent))
+        .collect())
```

Comment:
```suggestion
    let all_local_agents = local_agent_store
        .get_all()
        .await?
        .into_iter()
        .filter_map(|a| {
            if all_local_agents.contains(&a.agent().agent) {
                None
            } else {
                Some(a.agent.clone())
            }
        })
        .collect::<HashSet<_>>();
```

Rather than creating 2 vecs